src/nb_ndarray.cpp

```diff
@@ -335,9 +335,19 @@ ndarray_handle *ndarray_import(PyObject *o, const ndarray_req *req,
                 // c-style strides assumed
                 pass_order = req->req_order == 'C';
             } else {
+                bool size_is_zero = std::any_of(t.shape, t.shape + t.ndim,
+                                                [](auto &&shape) {return shape == 0;});
                 for (uint32_t i = 0; i < (uint32_t) t.ndim; ++i) {
-                    if (!((strides[i] == t.strides[i]) ||
-                          (t.shape[i] == 1 && t.strides[i] == 0))) {
+                    if (// regular case: strides should match
+                        !(strides[i] == t.strides[i] ||
+                          // if the shape of this dimension is 1, any
+                          // value of t.strides[i] is acceptable,
+                          // since 0 is the only valid index
+                          t.shape[i] <= 1 ||
+                          // if the shape of this dimension or any
+                          // faster-changing dimension is 0, any value of
+                          // t.strides[i] is acceptable
+                          size_is_zero)) {
```

Comment:
Rather than including this as part of the loop condition, I would make the entire loop conditional on `!size_is_zero`. I am not a fan of comments within expressions, could you move them to the top? 

Also, note that your comment states ` if the shape of this dimension is 1,` but the actual test is `<= 1` which includes zero. Since that case is already covered by the `size_is_zero` condition, you could change it to `== 1` to remove potential confusion caused by the overlap.